sql: validate function volatility

chengxiong-ruan opened the pull request:
This commits adds a simple udf volatility check by utilizing
the volatilities collected by the optbuilder when building a
statement. Errors are throw if any statement is not as strict
as the target volatility. Stable function folding is turned off
when building statements in CREATE FUNCTION to catch stable
function's volatility.

Release note (sql change): Previously, UDFs can be created with
any volatility no matter if the function body statements contain
any expression which would violate the target volatility. For
example, an immutable function might use `random()` in it. This
change added validations to guarantee that all statements in the
function body should be as strict as the expected UDF volatility.

Informs: #87699

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96476)
<!-- Reviewable:end -->

chengxiong-ruan committed:
sql: udf volatility check

This commits adds a simple udf volatility check by utilizing
the volatilities collected by the optbuilder when building a
statement. Errors are throw if any statement is not as strict
as the target volatility.

Release note (sql change): Previously, UDFs can be created with
any volatility no matter if the function body statements contain
any expression which would violate the target volatility. For
example, an immutable function might use `random()` in it. This
change added validations to guarantee that all statements in the
function body should be as strict as the expected UDF volatility.

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 26 of 26 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96476#-:-NNRG5xo5k21MA-5kXtK:bcmw7ss)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @mgartner)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_volatility_check` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96476#-NNRE4Ro5eo6w6hzDedD:-NNRE4Ro5eo6w6hzDedE:b-af0w6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9b9ec4beefe3aa1255aad55083baf2bcd0ebc99/pkg/sql/logictest/testdata/logic_test/udf_volatility_check#L97)):*
> ```
> ALTER FUNCTION f STABLE
> 
> statement error pq: volatile statement not allowed in immutable function: SELECT random\(\)::INT8
> ```

[nit] Do we have tests somewhere where the volatility is successfully altered in the other direction?


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/logictest/testdata/logic_test/udf_volatility_check` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96476#-NNRE4Ro5eo6w6hzDedD:-NNb3tKr9_qt9T8iFjuU:b-mjbc6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9b9ec4beefe3aa1255aad55083baf2bcd0ebc99/pkg/sql/logictest/testdata/logic_test/udf_volatility_check#L97)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Do we have tests somewhere where the volatility is successfully altered in the other direction?
</blockquote></details>

yes, tests are on line 77 and 88 where volatility are relaxed.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
TFTR!
bors r+
